composer/datasets/in_context_learning_evaluation.py

```diff
@@ -103,117 +108,171 @@ def _get_fewshot_sample_idxs(dataset_size: int, num_fewshot: int, sample_idx: in
+        self.encoded_dataset = self.dataset.map(
+            self._prep_example,
+            with_indices=True,
+            fn_kwargs={
+                'num_fewshot': num_fewshot,
+                'prompt_string': prompt_string,
+                'fewshot_rng': fewshot_rng,
+            },
+        )
 
-    def __init__(
+    def __getitem__(self, index: int):
+        return self.encoded_dataset[index]
+
+    def __len__(self):
+        return len(self.encoded_dataset)
+
+    def get_num_samples_in_batch(self, batch: dict) -> int:
+        return batch['input_ids'].shape[0]
+
+    def _read_dataset(
         self,
         dataset_uri: str,
-        tokenizer: Union[transformers.PreTrainedTokenizer, transformers.PreTrainedTokenizerFast],
-        max_seq_len: int,
-        pad_tok_id: int,
-        num_fewshot: int,
-        prompt_string: str,
-        example_delimiter: str,
-        continuation_delimiter: str,
         destination_path: str,
-        question_prelimiter: str,
-        fewshot_random_seed: int,
-        cot_delimiter: str = '',
+        hf_loading_vars: dict = None,
+        hf_parsing_vars: dict = None,
     ):
         try:
             from datasets import load_dataset  # pyright: ignore [reportGeneralTypeIssues]
         except ImportError as e:
-            raise MissingConditionalImportError(extra_deps_group='nlp',
-                                                conda_package='datasets',
-                                                conda_channel='conda-forge') from e
-        with dist.local_rank_zero_download_and_wait(destination_path):
-            if dist.get_local_rank() == 0:
-                get_file(dataset_uri, destination_path, overwrite=True)
-        dataset = load_dataset('json', data_files=destination_path, split='train', streaming=False)
-        self.samples = self._read_dataset(dataset)
-        self.samples = strip_data(self.samples)
-        self.tokenizer = tokenizer
-        self.max_seq_len = max_seq_len
-        self.pad_tok_id = pad_tok_id
-        self.padding_side = 'left'
-        self.max_answer_length = 0
-        fewshot_rng = random.Random(fewshot_random_seed)
-        self.encoded_dataset = self._prep_examples(num_fewshot, prompt_string, example_delimiter,
-                                                   continuation_delimiter, question_prelimiter, fewshot_rng,
-                                                   cot_delimiter)
-
-    def _format_prompt_and_fewshot(self, num_fewshot: int, prompt_string: str, example_delimiter: str,
-                                   continuation_delimiter: str, question_prelimiter: str, cot_delimiter: str,
-                                   fewshot_rng: random.Random, sample_idx: int) -> str:
+            raise MissingConditionalImportError(
+                extra_deps_group='nlp',
+                conda_package='datasets',
+                conda_channel='conda-forge',
+            ) from e
+        if 'hf://' in dataset_uri:
```

Comment:
probably check `startswith` instead of `in`?

Reply:
I don't really like this was to specify it's a HF dataset. It's different than the way we do HF models, where you just pass in the HF link. I'll look to see if there's a better way to differentiate btwn a local file, a cloud link, and a hf link.

Reply:
yeah easy enough:
(1) check os.path.exists for local file
(2) parse_uri to see if its remote
(3) then try hf
(4) error message explaining all tries failed

Reply:
I'd really prefer using hf:// as a prefix since that's how composer in general uses remote storage endpoints for files. What if there's somehow ambiguity about a local file that also exists? There's no need to allow such ambiguity

Reply:
fixed, I think just `hf://` works fine for now (with `.startswith()`)